internal/watchers/watcher.go

```diff
@@ -0,0 +1,206 @@
+// Copyright 2023 MongoDB Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package watchers
+
+import (
+	"errors"
+	"fmt"
+	"time"
+
+	atlasv2 "go.mongodb.org/atlas-sdk/admin"
+	atlas "go.mongodb.org/atlas/mongodbatlas"
+)
+
+type InvalidStateError struct {
+	State string
+}
+
+func (err *InvalidStateError) Error() string {
+	return fmt.Sprintf("Invalid state reached: %s", err.State)
+}
+
+type StateDescriber interface {
+	GetState() (string, error)
+}
+
+type Watcher struct {
+	Timeout            time.Duration // TODO: CLOUDP-181597
+	ExponentialBackoff bool
+	stateTransition    StateTransition
+	describer          StateDescriber
+	hasStarted         bool
+}
+
+const defaultWait = 4 * time.Second
+
+func NewWatcher(stateTransition StateTransition, describer StateDescriber) *Watcher {
+	return &Watcher{
+		stateTransition: stateTransition,
+		describer:       describer,
+	}
+}
+
+func (watcher *Watcher) Watch() error {
+	watcher.hasStarted = false
+	if watcher.ExponentialBackoff {
+		return watcher.exponentialBackoff()
+	}
+
+	return watcher.linearBackoff()
+}
+
+func (watcher *Watcher) exponentialBackoff() error {
+	backoff := defaultWait
+	for {
+		done, err := watcher.IsDone()
+		if err != nil || done {
+			return err
+		}
+		time.Sleep(backoff)
+		backoff *= 2
```

Comment:
typical backoff algorithm adjusts that value. For example many use fibonaci numbers for backoff. 
If we want to avoid complexity I would advise to:

- Remove backoff completely and use constant time
- Have backoff modeled by math.ceil(backoff*1.2) - that will give the best experience for Atlas API

Reply:
Could change it to fibonacci in this case, I don't have a strong opinion on this. I copied the behaviour of the exponential backoff we already have for the watch commands: https://github.com/mongodb/mongodb-atlas-cli/blob/9990292caf4a5509444e9da7f39e57ab453e8a5b/internal/cli/watch_opts.go#L61

Reply:
No strong opinion as well. Feel free to pick whatever works for you